swiftide/src/integrations/treesitter/summarizer.rs

```diff
@@ -0,0 +1,297 @@
+
+#[derive(Debug, Builder, Clone)]
+/// Generates a summary of a code file.
+///
+/// Supports splitting code files into chunks based on a maximum size or a range of bytes.
+#[builder(setter(into), build_fn(error = "anyhow::Error"))]
+pub struct CodeSummarizer {
+    #[builder(setter(custom))]
+    language: SupportedLanguages,
+}
+
+impl CodeSummarizerBuilder {
+    /// Attempts to set the language for the `CodeSummarizer`.
+    ///
+    /// # Arguments
+    ///
+    /// * `language` - A value that can be converted into `SupportedLanguages`.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<Self>` - The builder instance with the language set, or an error if the language is not supported.
+    ///
+    /// # Errors
+    /// * If the language is not supported, an error is returned.
+    pub fn try_language(mut self, language: impl TryInto<SupportedLanguages>) -> Result<Self> {
+        self.language = Some(
+            language
+                .try_into()
+                .ok()
+                .context("Treesitter language not supported")?,
+        );
+        Ok(self)
+    }
+}
+
+impl CodeSummarizer {
+    /// Creates a new `CodeSummarizer` with the specified language
+    ///
+    /// # Arguments
+    ///
+    /// * `language` - The programming language for which the code will be summarized.
+    ///
+    /// # Returns
+    ///
+    /// * `Self` - A new instance of `CodeSummarizer`.
+    pub fn new(language: SupportedLanguages) -> Self {
+        Self { language }
+    }
+
+    /// Creates a new builder for `CodeSummarizer`.
+    ///
+    /// # Returns
+    ///
+    /// * `CodeSummarizerBuilder` - A new builder instance for `CodeSummarizer`.
+    pub fn builder() -> CodeSummarizerBuilder {
+        CodeSummarizerBuilder::default()
+    }
+
+    /// Summarizes a code file.
+    ///
+    /// # Arguments
+    ///
+    /// * `code` - The source code to be split.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<String>` - A result containing a string, or an error if the code could not be parsed.
+    ///
+    /// # Errors
+    /// * If the code could not be parsed, an error is returned.
+    pub fn summarize(&self, code: &str) -> Result<String> {
+        let mut parser = Parser::new();
+        parser.set_language(&self.language.into())?;
+        let tree = parser.parse(code, None).context("No nodes found")?;
+        let root_node = tree.root_node();
+
+        if root_node.has_error() {
+            anyhow::bail!("Root node has invalid syntax");
+        }
+
+        let mut cursor = root_node.walk();
+        let mut summary = String::with_capacity(code.len());
+        let mut last_end = 0;
+        self.summarize_node(&mut cursor, code, &mut summary, &mut last_end);
+        Ok(summary)
+    }
+
+    fn is_unneeded_node(&self, node: Node) -> bool {
+        // We can use self.language to determine if a node is needed
+        match self.language {
+            SupportedLanguages::Rust => matches!(node.kind(), "block"),
```

Comment:
For my understanding, what is a block in tree-sitter grammar? I'm not sure if it's summarizing or doing something else.

Reply:
It's the list of statements that follows the function signature in Rust. So "summarization" in this file basically means removing all function bodies. Maybe there's a better word than "summarization".

Reply:
Oh yeah outline is a nice word, should we go with that instead?